ngraph/core/reference/src/runtime/reference/einsum.cpp

```diff
@@ -0,0 +1,1239 @@
+                    {
+                        new_shape.insert(new_shape.end(),
+                                         separate_sub_shape_prod.begin(),
+                                         separate_sub_shape_prod.end());
+                        new_shape.insert(new_shape.end(),
+                                         reduced_sub_shape_prod.begin(),
+                                         reduced_sub_shape_prod.end());
+                    }
+                    else
+                    {
+                        new_shape.insert(new_shape.end(),
+                                         reduced_sub_shape_prod.begin(),
+                                         reduced_sub_shape_prod.end());
+                        new_shape.insert(new_shape.end(),
+                                         separate_sub_shape_prod.begin(),
+                                         separate_sub_shape_prod.end());
+                    }
+
+                    // when new shape is equal to the current one,
+                    // there is no need in reshape
+                    if (new_shape == input->get_shape())
+                    {
+                        return input;
+                    }
+
+                    const auto element_type = input->get_element_type();
+                    const auto input_shape = input->get_shape();
+                    HostTensorPtr output =
+                        std::shared_ptr<HostTensor>(new HostTensor(element_type, new_shape));
+                    const AxisVector order = get_default_order(input_shape);
+
+                    ngraph::runtime::reference::reshape(
+                        reinterpret_cast<const char*>(input->get_data_ptr<T>()),
+                        reinterpret_cast<char*>(output->get_data_ptr<T>()),
+                        input_shape,
+                        order,
+                        new_shape,
+                        element_type.size());
+                    return output;
+                }
+
+                /// \brief      Contract two inputs of Einsum operation according to equation.
+                /// The result of the contraction is appended into inputs along with its
+                /// subscript. The inputs with indices input_ind1 and input_ind2 are removed from
+                /// inputs along with their input subscripts
+                ///
+                template <typename T>
+                void contract_two_inputs(HostTensorVector& inputs,
+                                         std::vector<std::string>& input_subscripts,
+                                         const std::string& output_subscript,
+                                         size_t input_ind1,
+                                         size_t input_ind2)
+                {
+                    // assume that input_ind1 < input_ind2 without loss of generality, otherwise,
+                    // just swap them
+                    if (input_ind2 < input_ind1)
```

Comment:
Looks like `input_ind1` is always `0`

Reply:
Now yes, but someday it can be changed once `compute_einsum_path` is complete.